### PR TITLE
[Tunnel] init module

### PR DESCRIPTION
**Pull request by satawatnack**
## Implementation details
Create Tunnel module

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: proto/band/tunnel/v1beta1/tunnel.proto [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+// TSSRoute is the type for a TSS route
+message TSSRoute {
+  option (cosmos_proto.implements_interface) = "RouteI";
+
+  // destination_chain_id is the destination chain ID
+  string destination_chain_id = 1 [(gogoproto.customname) = "DestinationChainID"];
+  // destination_contract_address is the destination contract address
+  string destination_contract_address = 2;
+}
```
Comment (RogerKSI) [inline proto/band/tunnel/v1beta1/tunnel.proto:23]: I think we should separate XXXRoute and XXXPacketContent into separate files like route.proto. to separate between core logic and route type. wdyt? 

Anyway, I'm ok with both. can just move them to be near each other.

File: x/tunnel/README.md [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+    - [MsgTriggerTunnel](#msgtriggertunnel)
+    - [MsgDepositToTunnel](#msgdeposittotunnel)
+    - [MsgWithdrawFromTunnel](#msgwithdrawfromtunnel)
+  - [Events](#events)
+    - [Event: `create_tunnel`](#event-create_tunnel)
+    - [Event: `update_and_reset_tunnel`](#event-update_and_reset_tunnel)
+    - [Event: `activate`](#event-activate)
+    - [Event: `deactivate`](#event-deactivate)
+    - [Event: `trigger_tunnel`](#event-trigger_tunnel)
+    - [Event: `produce_packet_fail`](#event-produce_packet_fail)
```
Comment (RogerKSI) [inline x/tunnel/README.md:42]: some event seems missing. e.g. produce_packet_success

File: x/tunnel/README.md [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+To create an IBC tunnel, use the following CLI command:
+
+> **Note**: An example of the signalInfos-json-file can be found at scripts/tunnel/signal_deviations.json.
+
+```bash
+bandd tx tunnel create-tunnel ibc [channel-id] [encoder] [initial-deposit] [interval] [signalInfos-json-file]
+```
+
+#### TSS Route
+
```
Comment (RogerKSI) [inline x/tunnel/README.md:105]: add TODO for this section.

File: x/tunnel/README.md [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+The `x/tunnel` module contains the following parameters:
+
+```go
+type Params struct {
+    // MinDeposit is the minimum deposit required to create a tunnel.
+    MinDeposit sdk.Coins
+    // MinInterval is the minimum interval in seconds.
+    MinInterval uint64
+    // MinInterval is the minimum interval in seconds.
+    MinInterval uint64
```
Comment (RogerKSI) [inline x/tunnel/README.md:193]: wrong field?

File: proto/band/tunnel/v1beta1/genesis.proto [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  // params is all parameters of the module.
+  Params params = 1 [(gogoproto.nullable) = false];
+  // tunnel_count is the number of tunnels.
+  uint64 tunnel_count = 2;
+  // tunnels is the list of tunnels.
+  repeated Tunnel tunnels = 3 [(gogoproto.nullable) = false];
+  // deposits is the list of deposits.
+  repeated Deposit deposits = 4 [(gogoproto.nullable) = false];
+  // total_fees is the type for the total fees collected by the tunnel
+  TotalFees total_fees = 5 [(gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (satawatnack) [reply proto/band/tunnel/v1beta1/genesis.proto:22]: it is the way to store sdk.Coins. This is one of the example in cosmos SDK
```
type ValidatorOutstandingRewards struct {
	Rewards github_com_cosmos_cosmos_sdk_types.DecCoins `protobuf:"bytes,1,rep,name=rewards,proto3,castrepeated=github.com/cosmos/cosmos-sdk/types.DecCoins" json:"rewards"`
}

// set validator outstanding rewards
func (k Keeper) SetValidatorOutstandingRewards(ctx sdk.Context, val sdk.ValAddress, rewards types.ValidatorOutstandingRewards) {
	store := ctx.KVStore(k.storeKey)
	b := k.cdc.MustMarshal(&rewards)
	store.Set(types.GetValidatorOutstandingRewardsKey(val), b)
}
```

File: x/tunnel/keeper/grpc_query.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+func (q queryServer) Deposits(
+	c context.Context,
+	req *types.QueryDepositsRequest,
+) (*types.QueryDepositsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	var deposits []*types.Deposit
+	depositStore := prefix.NewStore(ctx.KVStore(q.k.storeKey), types.DepositsStoreKey(req.TunnelId))
+
+	pageRes, err := query.Paginate(depositStore, req.Pagination, func(key []byte, value []byte) error {
```
[the remark this answers is not in the document]
Comment (satawatnack) [reply x/tunnel/keeper/grpc_query.go:82]: We did't filter anything

File: x/tunnel/keeper/keeper_latest_prices.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+func (k Keeper) MustGetLatestPrices(ctx sdk.Context, tunnelID uint64) types.LatestPrices {
+	latestPrices, err := k.GetLatestPrices(ctx, tunnelID)
+	if err != nil {
+		panic(err)
+	}
+	return latestPrices
+}
+
+// GetAllLatestPrices gets all the latest prices from the store
+func (k Keeper) GetAllLatestPrices(ctx sdk.Context) []types.LatestPrices {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_latest_prices.go:39]: These functions unused?

File: x/tunnel/keeper/keeper_tunnel.go [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+	for ; iterator.Valid(); iterator.Next() {
+		var tunnel types.Tunnel
+		k.cdc.MustUnmarshal(iterator.Value(), &tunnel)
+		tunnels = append(tunnels, tunnel)
+	}
+	return tunnels
+}
+
+// ActiveTunnelID sets the active tunnel ID in the store
+func (k Keeper) ActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_tunnel.go:170]: Suggestion: Rename this function to `AddActiveTunnelID`

File: x/tunnel/keeper/keeper_tunnel.go [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+	ctx.KVStore(k.storeKey).Set(types.ActiveTunnelIDStoreKey(tunnelID), []byte{0x01})
+}
+
+// DeactivateTunnelID deactivates the tunnel ID in the store
+func (k Keeper) DeactivateTunnelID(ctx sdk.Context, tunnelID uint64) {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_tunnel.go:175]: RemoveActiveTunnelID

File: x/tunnel/keeper/keeper_packet.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+		return types.Packet{}, types.ErrPacketNotFound.Wrapf("tunnelID: %d, sequence: %d", tunnelID, sequence)
+	}
+
+	var packet types.Packet
+	k.cdc.MustUnmarshal(bz, &packet)
+	return packet, nil
+}
+
+// MustGetPacket retrieves a packet by its tunnel ID and packet ID and panics if the packet does not exist
+func (k Keeper) MustGetPacket(ctx sdk.Context, tunnelID uint64, sequence uint64) types.Packet {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_packet.go:50]: Not use?

File: x/tunnel/keeper/keeper_packet.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	k.SetPacket(ctx, packet)
+
+	return nil
+}
+
+// calculateDeviationBPS calculates the deviation between the old price and
+// the new price in basis points, i.e., |(newPrice - oldPrice)| * 10000 / oldPrice
+func calculateDeviationBPS(oldPrice, newPrice sdkmath.Int) sdkmath.Int {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_packet.go:245]: move this function to helper?

File: x/tunnel/keeper/keeper_packet.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+	if oldPrice.IsZero() {
+		return sdkmath.NewInt(math.MaxInt64)
+	}
+
+	return newPrice.Sub(oldPrice).Abs().MulRaw(10000).Quo(oldPrice)
+}
+
+// CreatePricesMap creates a map of prices with signal ID as the key
+func CreatePricesMap(prices []feedstypes.Price) map[string]feedstypes.Price {
```
Comment (taobun) [inline x/tunnel/keeper/keeper_packet.go:258]: this too

**Review by taobun — COMMENTED** [review]
Minor comments overall looks ok

File: x/tunnel/keeper/keeper_tunnel.go (hunk 8 above)
Comment (satawatnack) [reply x/tunnel/keeper/keeper_tunnel.go:170]: Could we use this?
```
// SetActiveTunnelID sets the active tunnel ID in the store
func (k Keeper) SetActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
	ctx.KVStore(k.storeKey).Set(types.ActiveTunnelIDStoreKey(tunnelID), []byte{0x01})
}

// DeleteActiveTunnelID deletes the active tunnel ID from the store
func (k Keeper) DeleteActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
	ctx.KVStore(k.storeKey).Delete(types.ActiveTunnelIDStoreKey(tunnelID))
}
```

File: x/tunnel/types/packet.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	if err != nil {
+		return err
+	}
+	p.Receipt = any
+
+	return nil
+}
+
+// GetReceiptValue returns the packet's receipt.
+func (p Packet) GetReceiptValue() (PacketReceiptI, error) {
```
Comment (taobun) [inline x/tunnel/types/packet.go:49]: This too

File: x/tunnel/types/latest_prices.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
+package types
+
+import (
+	fmt "fmt"
```
Comment (RogerKSI) [inline x/tunnel/types/latest_prices.go:4]: no rename

File: x/tunnel/types/latest_prices.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+	if l.TunnelID == 0 {
+		return fmt.Errorf("tunnel ID cannot be 0")
+	}
+
+	if l.LastInterval < 0 {
+		return fmt.Errorf("last interval cannot be negative")
+	}
+
+	return nil
+}
```
Comment (RogerKSI) [inline x/tunnel/types/latest_prices.go:33]: seems not use anymore?

File: x/tunnel/types/tunnel.go [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+	if interval < params.MinInterval || interval > params.MaxInterval {
+		return ErrIntervalOutOfRange.Wrapf(
+			"max %d, min %d, got %d",
+			params.MaxInterval,
+			params.MinInterval,
+			interval,
+		)
+	}
+	return nil
+}
```
Comment (RogerKSI) [inline x/tunnel/types/tunnel.go:112]: these functions should not be here?

File: x/tunnel/README.md [hunk 17]
```diff
@@ -0,0 +1,538 @@
+# `x/tunnel`
+
```
Comment (RogerKSI) [inline x/tunnel/README.md:2]: update readme to current spec.

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: x/tunnel/types/packet.go (hunk 13 above)
Comment (satawatnack) [reply x/tunnel/types/packet.go:49]: GetReceipt() is already has in func generated by proto

File: x/tunnel/types/tunnel.go (hunk 16 above)
Comment (satawatnack) [reply x/tunnel/types/tunnel.go:112]: any suggest on where should I put ValidateInterval?

File: x/tunnel/types/packet.go (hunk 13 above)
Comment (taobun) [reply x/tunnel/types/packet.go:49]: Oh my

**Review by taobun — COMMENTED** [review]
Should we add method `GetTunnelRoute` or `GetRouteValue` on `Tunnel` struct? that make tunnel user don't need to try get cache value and cast to interface everytime?

**Comment by satawatnack** [comment]
> Should we add method `GetTunnelRoute` or `GetRouteValue` on `Tunnel` struct? that make tunnel user don't need to try get cache value and cast to interface?

Done!

File: x/tunnel/keeper/keeper_packet.go [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/bandprotocol/chain/v3/pkg/ctxcache"
+	feedstypes "github.com/bandprotocol/chain/v3/x/feeds/types"
+	"github.com/bandprotocol/chain/v3/x/tunnel/types"
+)
+
+// DeductBaseFee deducts the base fee from fee payer's account.
+func (k Keeper) DeductBasePacketFee(ctx sdk.Context, feePayer sdk.AccAddress) error {
```
Comment (RogerKSI) [inline x/tunnel/keeper/keeper_packet.go:16]: mismatch name.

File: x/tunnel/keeper/keeper_packet.go [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+	basePacketFee := k.GetParams(ctx).BasePacketFee
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, feePayer, types.ModuleName, basePacketFee); err != nil {
+		return err
+	}
+
+	// update total fees
+	totalFees := k.GetTotalFees(ctx)
+	totalFees.TotalPacketFee = totalFees.TotalPacketFee.Add(basePacketFee...)
```
Comment (RogerKSI) [inline x/tunnel/keeper/keeper_packet.go:24]: Should it be TotalBasePacketFee, or is it intended to be all packet fee?

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: x/tunnel/keeper/keeper_packet.go (hunk 19 above)
Comment (satawatnack) [reply x/tunnel/keeper/keeper_packet.go]: use TotalBasePacketFee

**Review by taobun — APPROVED** [review]

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**